src/AppInstallerCLICore/Workflows/WorkflowBase.cpp

```diff
@@ -730,27 +731,52 @@ namespace AppInstaller::CLI::Workflow
                 // The only time we don't want to output a line is when filtering and no update is available.
                 if (updateAvailable || !m_onlyShowUpgrades)
                 {
-                    Utility::LocIndString availableVersion, sourceName;
+                    Utility::LocIndString availableVersion, sourceName, sourceIdentifier;
+                    Utility::LocIndString packageId = match.Package->GetProperty(PackageProperty::Id);
 
                     if (latestVersion)
                     {
+                        // Always show the source for correlated packages
+                        sourceName = latestVersion->GetProperty(PackageVersionProperty::SourceName);
+                        sourceIdentifier = latestVersion->GetProperty(PackageVersionProperty::SourceIdentifier);
+
                         if (updateAvailable)
                         {
                             availableVersion = latestVersion->GetProperty(PackageVersionProperty::Version);
-                            availableUpgradesCount++;
+                            if (context.Args.Contains(Execution::Args::Type::ListAll) || !packageIdsPrinted.count({ packageId, sourceIdentifier }))
+                            {
+                                // we should only add to the upgrade count if we actually showed the table entry.
+                                availableUpgradesCount++;
+                            }
                         }
-
-                        // Always show the source for correlated packages
-                        sourceName = latestVersion->GetProperty(PackageVersionProperty::SourceName);
                     }
 
-                    table.OutputLine({
-                        match.Package->GetProperty(PackageProperty::Name),
-                        match.Package->GetProperty(PackageProperty::Id),
-                        installedVersion->GetProperty(PackageVersionProperty::Version),
-                        availableVersion,
-                        shouldShowSource ? sourceName : ""s
-                        });
+                    if (context.Args.Contains(Execution::Args::Type::ListAll))
+                    {
+                       table.OutputLine({
+                           installedVersion->GetProperty(PackageVersionProperty::Name),
+                           match.Package->GetProperty(PackageProperty::Id),
+                           installedVersion->GetProperty(PackageVersionProperty::Version),
+                           availableVersion,
+                           shouldShowSource ? sourceName : ""s
+                       });
+                    }
+                    else
+                    {
+                        // we need to only list once per package
+                        if (!packageIdsPrinted.count({ packageId, sourceIdentifier }))
+                        {
+                            packageIdsPrinted.insert({ packageId, sourceIdentifier });

```

Comment:
```suggestion
                            packageIdsPrinted.emplace(std::move(thePair));
```
Assuming that you also took my other suggestion above about keeping these as a `pair` rather than separate.  And you named it `thePair`, which I'm sure you would come up with a better name than that 😉